### ISSUE TITLE
Rule: DI preferences

### ISSUE THREAD

**Issue by jissereitsma** [extdn/extdn-phpcs#1]
Report when `di.xml` is overwriting preferences that originate in the Magento core framework or core modules.

**Comment by avstudnitz** [comment]
This might be needed and not be harmful in some cases. We all know that the Magento 2 core is not perfect, and especially the older parts of the code with long methods and lots of logic in protected methods cannot be changed nicely with plugins.

**Comment by jissereitsma** [comment]
Correct. As I see it, we will have two types of uses for these PHPCS rules: First, a mandatory set of rules to which all should comply. Second, an "optional" set of rules that reports certain things. A preference overriding a Magento core class is something that definitely needs reporting in some way. Unfortunately, PHPCS only seems to have calls like `addWarning()` and `addError()` and not something like `addNotice()`. Maybe a warning should be picked here?
